proto/src/serializers/bech32str.rs

```diff
@@ -0,0 +1,96 @@
+//! Bech32 serializers. Because Bech32 is parameterized by the HRP, this module
+//! implements (internal) helper functions that are used in submodules that fill
+//! in parameters for various key types.
+
+use bech32::{FromBase32, ToBase32};
+// re-exporting allows the convenience methods to be used without referencing
+// the underlying bech32 crate.
+pub use bech32::{Variant, Variant::*};
+use serde::{Deserialize, Deserializer, Serializer};
+
+/// Convenience method for (general-purpose) Bech32 decoding.
+///
+/// Works around a bit of awkwardness in the [`bech32`] API.
+pub fn decode(
+    string: &str,
+    expected_hrp: &str,
+    expected_variant: Variant,
+) -> anyhow::Result<Vec<u8>> {
+    let (hrp, data, variant) = bech32::decode(&string)?;
+
+    if variant != expected_variant {
+        return Err(anyhow::anyhow!(
+            "wrong bech32 variant {:?}, expected {:?}",
+            variant,
+            expected_variant
+        ));
+    }
+    if hrp != expected_hrp {
+        return Err(anyhow::anyhow!(
+            "wrong bech32 human readable part {}, expected {}",
+            hrp,
+            expected_hrp
+        ));
+    }
+
+    Ok(Vec::from_base32(&data).expect("bech32 decoding produces valid base32"))
+}
```

Comment:
The motivation for having separate, public `decode` and `encode` methods is to avoid having to do this dance every time someone wants to do manual Bech32 serialization.

The reason they'd want to do that is in order to implement `Display`/`FromStr`, the type needs to be able to pass to and from Bech32 directly, and I'm not sure that there's a way to reuse the Serde implementation to do so. (unless?)